src/polydatum/middleware.py

```diff
@@ -0,0 +1,184 @@
+
+    For example, in this code:
+
+        dal.my_service.sub_service.other_service.method()
+
+    For the `dal` object, the entire attribute path would be:
+
+        my_service.sub_service.other_service.method
+
+    Which may be made up of PathSegments like:
+
+        PathSegment(name="my_service"), PathSegment(name="sub_service"), etc.
+    """
+    name = None
+
+    # A collection of meta properties.
+    # Example:
+    #   meta = {
+    #       "enabled": True
+    #   }
+    meta = None
+
+    def __init__(self, name: str, **meta):
+        self.name = name
+        self.meta = meta
+
+    def __eq__(self, other):
+        return all([
+            isinstance(self, PathSegment),
+            self.name == other.name,
+            self.meta == other.meta
+        ])
+
+
+class DalMethodRequest:
+    """
+    Mutable request state. Middleware can modify this.
+    """
+
+    def __init__(
+            self, ctx: DataAccessContext, path: Tuple[PathSegment, ...], args: Tuple[Any, ...], kwargs: Dict
+    ):
+        self.ctx = ctx
+        self.path = path
+        self.args = args
+        self.kwargs = kwargs
+        self.dal_method = None  # Not resolved yet
+
+
+class DalMethodError(Exception):
+    """
+    This class defines a DalMethodError that is raised
+    when a service/method cannot be resolved by the
+    DalMethodResolverMiddleware
+    """
+    def __init__(self, path: Optional[Tuple[PathSegment]] = None):
+        self.path = path
+        super().__init__(f'Invalid dal method: {self.path}')
+
+
+class DeferredDalAttrRequest:
+    """
+    A utility class that defers access to DAL attributes.
+
+    Attribute access to the DAL is deferred so that the DAL can
+    run method middleware before and after calling a method.
+
+    Method middlewares may prevent method's from being called
+    by raising exceptions.
+    """
+
+    # Defining type here allows subclasses to easily provide another class
+    PathSegment = PathSegment
+
+    def __init__(self, handler: Callable, path: Tuple[PathSegment, ...]):
+        """
+        Args:
+            handler (Callable): A callable that handles calling the underlying
+                attribute by wrapping it so that method middleware can run before
+                and after calling a method
+            path (Tuple[PathSegment, ...]): A tuple of PathSegments representing
+                the current attribute path that has been requested.
+        """
+        self._handler = handler
+        self.path = path
+
+    def __getattr__(self, name: str) -> 'DeferredDalAttrRequest':
+        return self.__class__(self._handler, self.path + (self.PathSegment(name=name),))
+
+    def __call__(self, *args, **kwargs):
+        return self._handler(self.path, *args, **kwargs)
+
+    def __repr__(self):
+        location = [p.name for p in self.path]
+        location = '.'.join(location)
+        return f"<{self.__class__} paths={location}>"
```

Comment:
`__repr__` is supposed to return something you could paste into a Python REPL and it would instantiate. This uses a string for `paths` instead of a tuple of `PathSegment` so it would fail.

`__str__` has no such conventions. You can return whatever.